packages/core/src/config-manager/service-discovery/index.ts

```diff
@@ -1,44 +1,70 @@
-import { ModuleListResponse, RegisteredModule } from '@conduitplatform/commons';
+import {
+  DeploymentState,
+  DeploymentState_ModuleStateInfo as ModuleStateInfo,
+  Empty,
+  ModuleHealthRequest,
+  RegisterModuleRequest,
+  RegisterModuleRequest_ConduitManifest as ConduitManifest,
+} from '@conduitplatform/commons';
 import ConduitGrpcSdk, {
-  GrpcCallback,
   GrpcRequest,
   GrpcResponse,
   HealthCheckStatus,
   linearBackoffTimeout,
+  ManifestManager,
+  ModuleActivationStatus,
 } from '@conduitplatform/grpc-sdk';
-import { IModuleConfig } from '../../interfaces/IModuleConfig';
+import {
+  ModuleClient,
+  ModuleRegistrationState,
+  ServiceDiscoveryState,
+} from '../../interfaces';
 import { ServerWritableStream, status } from '@grpc/grpc-js';
 import { EventEmitter } from 'events';
 import { clearTimeout } from 'timers';
 
-/*
+/**
  * - Multi-instance services are not handled individually (LoadBalancer)
+ * - Services are distinguished between Available and Pending based on whether their dependencies are met
  * - Online Services are recovered on startup
  * - Unresponsive services are instantly removed from the list of exposed services
  * - Reconnection to recently removed services is attempted using linear backoff
  * - Services that do not provide a gRPC health check service are assumed to be healthy
  */
 export class ServiceDiscovery {
-  readonly registeredModules: Map<string, RegisteredModule> = new Map<
-    string,
-    RegisteredModule
-  >();
+  readonly registeredModules: Map<string, ModuleStateInfo>;
+  private readonly moduleManifests: Map<string, ConduitManifest> = new Map();
   private readonly moduleHealth: {
     [module: string]: { address: string; timestamp: number; status: HealthCheckStatus };
   } = {};
+  private deploymentStateModified: boolean = false;
   private readonly moduleRegister: EventEmitter;
-  private servingStatusUpdate: boolean = false;
   private monitorIntervalMs = 30000;
   private serviceReconnRetries = 5;
   private serviceReconnInitMs = 250;
 
-  constructor(private readonly grpcSdk: ConduitGrpcSdk) {
+  constructor(
+    private readonly grpcSdk: ConduitGrpcSdk,
+    private readonly conduitHealthCheckGetter: (service: '') => HealthCheckStatus,
+  ) {
     this.moduleRegister = new EventEmitter();
     this.moduleRegister.setMaxListeners(150);
+    this.registeredModules = new Map([
+      [
+        'conduit',
+        {
+          moduleName: 'conduit',
+          moduleVersion: ManifestManager.getInstance().moduleVersion,
+          moduleUrl: 'unknown', // TODO
```

Comment:
This doesn't technically break anything, but it's still an eye-sore that should probably be addressed.